tools/clang/lib/CodeGen/CGExprCXX.cpp

```diff
@@ -235,12 +235,17 @@ RValue CodeGenFunction::EmitCXXMemberOrOperatorMemberCallExpr(
 
           llvm::Constant *zero = Builder.getInt32(0);
           llvm::Value *TmpThis = CreateTempAlloca(Ty);
+          QualType ElTy = hlsl::GetElementTypeOrType(Base->getType());
+          bool IsBool = ElTy->isSpecificBuiltinType(BuiltinType::Bool);
           for (unsigned i = 0; i < Ty->getVectorNumElements(); i++) {
             llvm::Value *EltIdx = Elts->getAggregateElement(i);
             llvm::Value *EltGEP = Builder.CreateGEP(This, {zero, EltIdx});
             llvm::Value *TmpEltGEP =
                 Builder.CreateGEP(TmpThis, {zero, Builder.getInt32(i)});
             llvm::Value *Elt = Builder.CreateLoad(EltGEP);
+            if (IsBool)
+              Elt = Builder.CreateTrunc(
+                  Elt, llvm::Type::getInt1Ty(getLLVMContext()));
```

Comment:
Will this impact conversion of bool that could come from host memory?  Trunc is only valid if you know that the source is already a bit pattern constrained by the compiler.  Otherwise, we must use comparison with zero (not-equal) to translate from i32 in storage to i1.

Reply:
It is undefined behavior in C/C++ and HLSL for bool to have any value other than 0 or 1. Given that, trunc is valid in all cases converting between lengths of bool.

Reply:
I've applied Chris' fixes based on your comments @tex3d. I'm waiting for your approval before merging.

Reply:
To fully cite the spec. This is the wording in the C++ spec:

> Values of type bool are either true or false.
[HTML Citation](https://timsong-cpp.github.io/cppwp/n4140/basic.fundamental#6)

Since HLSL defines `true` to `1` and `false` to `0` (the same as pretty much all C/C++ compilers), any different value is UB.

Reply:
While I can't find documentation to back this up now, HLSL compilers (FXC and DXC) have always required a translation for bools coming from host visible memory (cbuffer, etc...) by interpreting any non-zero value as true, and zero value as false.  It's something we did in FXC, and had to do in DXC to match FXC.  Trunc will convert any bit pattern with the lowest bit unset to `0`.  This would be inconsistent with the way bools coming from host memory have been handled thus far for HLSL.

Reply:
For one thing, DXBC instructions are defined in a way that an incoming bool value from a constant buffer may be used in an `if` instruction, and the `if` instruction interprets the 32-bit value with any bit set as `true`, and no bits set as `false`.
See the instruction spec [here](https://microsoft.github.io/DirectX-Specs/d3d/archive/D3D11_3_FunctionalSpec.htm#22.7.1%20Branch%20based%20on%20boolean%20condition:%20if_condition).

Reply:
Though, as luck would have it, under `Restrictions:` `2)` it gets `_z` and `_nz` modifiers exactly backwards.  The earlier description of the operation gets it right though.

Reply:
I don’t see how the behavior of the DXBC instruction is relevant here. This is absolutely not a behavior you can rely on LLVM to match. Clang (and by extension DXC) have been assuming that the only valid values for bool are 0 and 1 (see the bit where this PR changes the range check).

If a bool value comes in from somewhere and it is anything other than 0 or 1, it’s going to break in a lot of places. The advantage to using a `trunc` here is that it is _way_ easier for LLVM to optimize and eliminate unnecessary `trunc`, `zext` and `sext` instructions because the don’t always depend on value analysis.

If you feel strongly enough about this to hold up the PR we should have a meeting to discuss.

Reply:
We had a lengthy discussion internally and I think we all went back and forth about different thoughts here. In practice since the existing code causes a crash, and this causes correct behavior where the only possible error cases are very firmly in the UB, I think we should take the change as-is.

If @tex3d feels strongly about this we can discuss this more and revisit.

There are a lot of broken behaviors around HLSL boolean vectors that we're going to need to revisit and reconsider in our Clang implementation and we can discuss longer term solutions in that context.